documentation/ADRs/027_ensmeble_reconciliation.md

```diff
@@ -0,0 +1,47 @@
+Ensemble reconcilation
+
+| ADR Info            | Details                 |
+|---------------------|-------------------------|
+| Subject             | Ensemble reconciliation |
+| ADR Number          | 027                     |
+| Status              | proposed                |
+| Author              | Jim                     |
+| Date                | 01/11/2024              |
+
+## Context
+The notebook-based views3/fatalities002 pipeline generates a cm and a pgm ensemble. It was found that the pgm ensemble suffered from what might be termed normalisation issues, in that the peak and total numbers of fatalities forecast at pgm level are clearly too low. In particular, summing forecast fatalities over the pg cells belonging to a given country, a dcomapring to the fatalities forecast for the same country at cm level almost always gives the result that the summed pgm values are significantly - often an order of magnitude - lower.
+As a quick fix, therefore, a reconciliation function was created which accepts a pgm forecast dataframe and a cm forecast dataframe, fetches via viewser a pgm->cm mapping, computes for every country for every month the sum over its constituent pg cells, and renormalises the pgm forecasts for those cells so that the sum matches the cm-level forecast. A check is performed which ensures that the set of months in the two input dfs is the same.
+The reconciliation will be applied to every pgm-level constituent model from which the pgm ensemble is built.
+
+## Decision
+This reconciliation is to be implemented in the pipeline as a temporary fix in lieu of improvements to the pgm models. The reconciliation function itself needs to be globally available, so should live in common utils.
+For each ensemble, a new item of metadata will be created, 'reconcile_with', whose value will either be None, or the name of another ensemble. No checks need be done on whether a valid choice has been made, since the function already checks to see that the two ensembles it is presented with have correctly-formatted indexes, and the identical month-sets. This change needs to be present in the ensemble-creation meta-tool, with the default value of None.
+In an ensemble's generate_forecast.py, a code fragment needs to be added where if reconcile_with is not None, the ensemble named by reconcile_with is fetched from storage and presented to the reconciliation function along with each pgm constituent model in turn. 
+Warnings are to be issued and logged if negative-valued forecasts are encountered (before setting them to zero) and if large normalisations are necessary.
```

Comment:
Could we get something like this (or is that a  different ADR?)

Here’s the streamlined table as requested:

---

| Aspect                | Condition                          | Action                              | Threshold/Details                              | Purpose                                                      |
|-----------------------|------------------------------------|-------------------------------------|------------------------------------------------|--------------------------------------------------------------|
| **Negative Values**   | Any forecast < 0                   | Set to zero; log a warning         | Individual negative values                      | Ensures non-negative outputs for models where negatives are invalid |
| **Frequent Negatives**| >10% of forecasts in a batch < 0  | Trigger a critical alert           | More than 10% negative values in a single batch| Indicates severe model issues that may require immediate attention or retraining |
| **Moderate Adjustments** | Adjustment >10% but ≤50%        | Log a warning                       | Adjustments between 10% and 50%                | Signals a potential drift or input anomaly; enables early detection |
| **Extreme Adjustments** | Adjustment >50%                  | Trigger a critical alert           | Adjustments greater than 50%                   | Alerts the team to significant discrepancies requiring urgent review |
| **Traceability**      | N/A                                | Include model, ensemble, timestamp | Applies to all logs and warnings               | Ensures logs provide enough context for debugging and root cause analysis |

---

Reply:
This should be another ADR I think

Reply:
Super